File: UnitB/Exercises/unitb-ex-28.jl
include("./init.jl")
using Bootstrap,UnicodePlots

desc=Lock5Table(386,"SkateboardPrices","Skateboard Prices?",["Price"])
data=@pipe load_csv(desc.name)

##  3. question

### 3.1 mean and std
  "mean_std"=>mean_and_std(data.Price)
  # "mean_std" => (67.58649999999999, 50.018024253894296)

### 3.2 bootstrap sampling
   n_boot=1000
   bs(data) = bootstrap(mean,data, BasicSampling(n_boot))
   bs1=bs(data.Price)
   #= 
     Bootstrap Sampling
  Estimates:
     Var │ Estimate  Bias      StdError
         │ Float64   Float64   Float64
    ─────┼──────────────────────────────
       1 │  67.5865  0.085111   11.1013
  Sampling: BasicSampling
  Samples:  1000
  Data:     Vector{Float64}: { 20 }
   =#


### 3.3  frequency

   histogram(bs1.t1[1])

   #= 
     
                  ┌                                        ┐ 
   [ 30.0,  40.0) ┤▎ 1                                       
   [ 40.0,  50.0) ┤███▊ 39                                   
   [ 50.0,  60.0) ┤██████████████████████▎ 225               
   [ 60.0,  70.0) ┤███████████████████████████████████  354  
   [ 70.0,  80.0) ┤█████████████████████████▎ 254            
   [ 80.0,  90.0) ┤██████████▎ 103                           
   [ 90.0, 100.0) ┤██▏ 21                                    
   [100.0, 110.0) ┤▍ 3                                       
                  └                                        ┘ 
                                   Frequency                 
   =#

### 3.4 confidence interval
   cil = 0.95;
   ## basic CI
   bci1 = confint(bs1, BasicConfInt(cil))
   # ((67.58649999999999, 44.37641249999997, 86.66351249999997),)

   

File: marimo/Exercise/unitb-ex-28.py
import marimo

__generated_with = "0.24.0"
app = marimo.App(width="medium")


@app.cell
def _():
    import marimo as mo
    import sys
    from pathlib import Path
    import numpy as np
    import polars as pl
    import matplotlib.pyplot as plt
    from scipy import stats
    from great_tables import GT
    _p = Path(mo.notebook_location()).resolve().parent
    while _p != _p.parent and not (_p / "lock5.py").exists():
        _p = _p.parent
    if str(_p) not in sys.path:
        sys.path.insert(0, str(_p))
    import lock5
    return mo, sys, Path, np, pl, plt, stats, GT, lock5


@app.cell
def _(mo):
    mo.md(r"""# Unit B Ex 28 — skateboard prices bootstrap

Port of `UnitB/Exercises/unitb-ex-28.jl`. Bootstrap the mean `Price` (1000
samples), report bias / std error / basic 95% CI, and histogram the bootstrap
means. Bootstrap results are random, so values differ from the Julia run.""")
    return


@app.cell
def _(lock5):
    desc = lock5.Lock5Table(386, "SkateboardPrices", "Skateboard Prices?", ["Price"])
    data = lock5.load_csv(desc.name)
    prices = data["Price"].to_numpy().astype(float)
    return data, desc, prices


@app.cell
def _(lock5, mo, np, pl, prices):
    mn, sd = lock5.mean_and_std(prices)
    boots = lock5.boot_sampling(prices, n=1000, method=np.mean, seed=1)
    estimate = float(prices.mean())
    bci = lock5.basic_confint(boots, estimate, 0.95)
    res_df = pl.DataFrame({
        "item": ["mean", "std (sample)", "bias", "std error",
                 "basic 95% CI lower", "basic 95% CI upper"],
        "value": [round(mn, 4), round(sd, 4), round(float(boots.mean() - estimate), 4),
                  round(float(boots.std(ddof=1)), 4), round(bci[1], 4), round(bci[2], 4)],
    })
    mo.Html(lock5.gt_table(res_df, title="Skateboard prices bootstrap (1000 samples)").as_raw_html())
    return bci, boots, estimate, res_df


@app.cell
def _(boots, plt):
    fig, ax = plt.subplots(figsize=(7, 4))
    ax.hist(boots, bins=8, color="gray", alpha=0.6, edgecolor="black", linewidth=1)
    ax.set_title("Bootstrap mean Price")
    ax.set_xlabel("Price")
    ax.set_ylabel("Frequency")
    fig
    return fig, ax


if __name__ == "__main__":
    app.run()
